File: android-runner/sdk/src/main/kotlin/dev/smix/sdk/Modifiers.kt
// Modifiers data class + AnchorBox + IndexModifiers mirror the Swift
// SDK + Rust smix-selector struct shape.
//
// Wire JSON shape: Modifiers fields are FLATTENED into the Selector
// body via custom KSerializer (matches Rust `#[serde(flatten)]`).
// AnchorBox is the spatial-only subset for Selector.Anchor.
// IndexModifiers is the index-only subset.

package dev.smix.sdk

import kotlinx.serialization.Serializable

/**
 * All-optional modifier set stacked onto a base Selector.
 * Spatial fields hold sub-selectors; index fields are scalar picks.
 *
 * Serialization: fields are FLATTENED into the Selector JSON body
 * via SelectorSerializer; the standalone @Serializable is used only
 * for nested decoding helpers.
 */
@Serializable
data class Modifiers(
    val near: Selector? = null,
    val below: Selector? = null,
    val above: Selector? = null,
    val leftOf: Selector? = null,
    val rightOf: Selector? = null,
    val inside: Selector? = null,
    val ancestor: Selector? = null,
    val nth: Int? = null,
    val first: Boolean? = null,
    val last: Boolean? = null,
) {
    /** True when no modifier field is set. */
    val isEmpty: Boolean
        get() = near == null && below == null && above == null &&
            leftOf == null && rightOf == null && inside == null &&
            ancestor == null && nth == null && first == null && last == null

    companion object {
        val EMPTY = Modifiers()
    }
}

/**
 * Spatial-only anchor box (no index, no base). Used by
 * [Selector.Anchor] — at least one field must be non-null for the
 * resolver to produce candidates.
 */
@Serializable
data class AnchorBox(
    val near: Selector? = null,
    val below: Selector? = null,
    val above: Selector? = null,
    val leftOf: Selector? = null,
    val rightOf: Selector? = null,
    val inside: Selector? = null,
    val ancestor: Selector? = null,
)

/** Index-only modifier subset used by [Selector.Anchor]. */
@Serializable
data class IndexModifiers(
    val nth: Int? = null,
    val first: Boolean? = null,
    val last: Boolean? = null,
) {
    companion object {
        val EMPTY = IndexModifiers()
    }
}

File: swift-bridge/Sources/SmixSDK/Modifiers.swift
// Modifiers struct — mirrors Rust smix-selector `Modifiers` +
// `IndexModifiers`, flattened into the Selector body via custom Codable.
//
// Wire JSON shape: fields are FLATTENED into the Selector base body.
// E.g. Rust:
//   {"id": "btn-foo", "below": {"id": "anchor"}, "nth": 2}
// not:
//   {"id": "btn-foo", "modifiers": {"below": {...}, "nth": 2}}
//
// Each field is Optional and `nil` when omitted (skipped on encode).
//
// Spatial: near / below / above / leftOf / rightOf / inside / ancestor
// Index:   first / last / nth
//
// Stackable: e.g. `{"id":"btn", "below":{"text":"hi"}, "nth": 0, "near":{"role":"button"}}`.

import Foundation

/// All-optional modifier set stacked onto a base [`Selector`].
///
/// Constructed empty via `.empty` then mutated by fluent chaining
/// methods on Selector. Spatial fields hold sub-selectors (anchor
/// to spatially compare against); index fields are scalar picks.
public struct Modifiers: Sendable, Equatable, Codable {
    // ---- spatial -----------------------------------------------------

    /// Candidate must be geometrically near this anchor (centroid-distance
    /// ≤ threshold pts). Threshold defaults to 100pt per Rust resolver.
    public var near: Selector?
    /// Candidate must be geometrically below this anchor (larger y center).
    public var below: Selector?
    /// Candidate must be geometrically above this anchor (smaller y center).
    public var above: Selector?
    /// Candidate must be to the left of this anchor (smaller x center).
    public var leftOf: Selector?
    /// Candidate must be to the right of this anchor (larger x center).
    public var rightOf: Selector?
    /// Candidate must be geometrically inside this anchor's bounds.
    public var inside: Selector?
    /// Ancestor sub-selector — candidate's a11y tree parent chain (recursive
    /// ancestors, excluding self) must contain at least one node matching
    /// this sub-selector. Structural filter, not spatial.
    public var ancestor: Selector?

    // ---- index -------------------------------------------------------

    /// Pick the nth match (0-indexed) from surviving candidates.
    public var nth: Int?
    /// Pick the first match (`true`). Overridden by `nth` if both set.
    public var first: Bool?
    /// Pick the last match (`true`). Overridden by `nth` if both set.
    public var last: Bool?

    public init(
        near: Selector? = nil,
        below: Selector? = nil,
        above: Selector? = nil,
        leftOf: Selector? = nil,
        rightOf: Selector? = nil,
        inside: Selector? = nil,
        ancestor: Selector? = nil,
        nth: Int? = nil,
        first: Bool? = nil,
        last: Bool? = nil
    ) {
        self.near = near
        self.below = below
        self.above = above
        self.leftOf = leftOf
        self.rightOf = rightOf
        self.inside = inside
        self.ancestor = ancestor
        self.nth = nth
        self.first = first
        self.last = last
    }

    /// Empty modifiers — no filtering or index pick applied.
    public static let empty = Modifiers()

    /// True when no modifier field is set — used by encoder to skip the
    /// shared modifier block entirely.
    public var isEmpty: Bool {
        near == nil && below == nil && above == nil
            && leftOf == nil && rightOf == nil && inside == nil
            && ancestor == nil && nth == nil && first == nil && last == nil
    }
}
